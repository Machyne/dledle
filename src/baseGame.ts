// Give these string values so they can be round trip through being used as object keys.
// These are never encoded as-is so strings are fine.
export enum GameScore {
  Win = "Win",
  NearWin = "NearWin",
  Loss = "Loss",
}

export type GameResult = {
  score: GameScore;
  serializedResult: string;
};

export abstract class BaseGame {
  private _resultRegex: RegExp | null = null;

  /**
   * The name of the game.
   */
  abstract get name(): string;

  /**
   * The name of the game in lowercase with only alphanumeric characters.
   */
  get fileName(): string {
    return this.name.toLowerCase().replace(/[^a-z0-9]/g, "");
  }

  /**
   * A regex that matches the result of the game.
   */
  get resultRegex(): RegExp {
    if (this._resultRegex === null) {
      this._resultRegex = this._buildResultRegex();
    }
    return this._resultRegex;
  }

  /**
   * Builds a regex that matches the result of the game. The return value will be cached.
   */
  abstract _buildResultRegex(): RegExp;

  /**
   * A link to the game.
   */
  abstract get link(): string;

  /**
   * Parses the user-supplied game result and returns a serialized version of
   * the game data in addition to the "score".
   *
   * @param gameResult the result of the game run through the resultRegex
   */
  abstract serializeResult(gameResult: RegExpMatchArray): GameResult;

  /**
   * The share text of the result of the game given the serialized game data.
   *
   * @param serializedResult the serialized game data
   */
  abstract deserialize(serializedResult: string): string;
}
