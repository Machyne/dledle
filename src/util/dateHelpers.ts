export type DateParts = {
  year: number;
  month: number;
  day: number;
};

export function ymd(year: number, month: number, day: number): DateParts {
  return { year, month, day };
}
export const FIRST_GAME_DATE: DateParts = ymd(2023, 12, 21);

/*
 * Returns the number of days between two dates (ie `a - b` in days).
 * If a is earlier than b, the result will be negative.
 */
export function signedDayDifference(a: DateParts, b: DateParts): number {
  const aDate = new Date(a.year, a.month - 1, a.day);
  const bDate = new Date(b.year, b.month - 1, b.day);
  return Math.round((aDate.getTime() - bDate.getTime()) / (1000 * 60 * 60 * 24));
}

export function gameNumberForDate(date: DateParts): number {
  // Add 1 because the first game is game 1, not game 0.
  return signedDayDifference(date, FIRST_GAME_DATE) + 1;
}

export function currentGameNumber(): number {
  return gameNumberForDate(currentOfficialDate());
}

export function currentOfficialDate(): DateParts {
  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone: "America/Chicago",
    year: "numeric",
    month: "numeric",
    day: "numeric",
  });
  const { year, month, day } = Object.fromEntries(
    formatter.formatToParts(new Date()).map(({ type, value }) => [type, parseInt(value)]),
  );
  return { year, month, day };
}
