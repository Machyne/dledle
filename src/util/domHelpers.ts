export function baseUrl() {
  let port = window.location.port;
  if (port) {
    port = ":" + port;
  }
  return window.location.protocol + "//" + window.location.hostname + port + "/";
}

export function el<T extends HTMLElement = HTMLElement>(id: string): T | null {
  const e = document.getElementById(id) as T | null;
  if (!e) {
    console.error(`No element with id ${id}`);
  }
  return e;
}

export function elOrThrow<T extends HTMLElement = HTMLElement>(id: string): T {
  const e = document.getElementById(id) as T | null;
  if (!e) {
    throw new Error(`No element with id ${id}`);
  }
  return e;
}

export function show(elementOrID: string | Element | null) {
  if (!elementOrID) {
    return;
  }
  if (typeof elementOrID === "string") {
    elementOrID = elOrThrow(elementOrID);
  }
  elementOrID.classList.remove("hidden");
}

export function hide(elementOrID: string | Element | null) {
  if (!elementOrID) {
    return;
  }
  if (typeof elementOrID === "string") {
    elementOrID = elOrThrow(elementOrID);
  }
  elementOrID.classList.add("hidden");
}

export function addHideListener(element: Element, callback: () => void) {
  let wasHidden = element.classList.contains("hidden");
  const attrObserver = new MutationObserver((mutations) => {
    mutations.forEach((mu) => {
      if (mu.type !== "attributes" && mu.attributeName !== "class") return;
      const currentlyHidden = element.classList.contains("hidden");
      if (!wasHidden && currentlyHidden) {
        callback();
      }
      wasHidden = currentlyHidden;
    });
  });
  attrObserver.observe(element, { attributes: true });
}
