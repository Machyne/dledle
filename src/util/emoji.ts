import Graphemer from "graphemer";

const splitter = new Graphemer();

export function splitEmoji(str: string): Array<string> {
  return splitter.splitGraphemes(str).filter((c) => !/^\s+$/u.test(c));
}

export function splitEmojiLines(str: string): Array<Array<string>> {
  const lines = str.split(/\n/u);
  const splitLines = lines.map((line) => splitEmoji(line));
  return splitLines.filter((line) => line.length > 0);
}

export function emojiToRegexUnion(emoji: Array<string>) {
  return `(?:${emoji.join("|")})`;
}

export const cardinalArrowString = "⬆️➡️⬇️⬅️";
export const ordinalArrowString = cardinalArrowString + "↗️↘️↙️↖️";
export const cardinalArrows = splitEmoji(cardinalArrowString);
export const ordinalArrows = splitEmoji(ordinalArrowString);
export const ordinalArrowValues: Record<string, number> = Object.fromEntries(
  ordinalArrows.map((key, idx) => [key, idx]),
);
export const ordinalArrowRegex = emojiToRegexUnion(ordinalArrows);
export const arrows = {
  u: "⬆️" as const,
  r: "➡️" as const,
  d: "⬇️" as const,
  l: "⬅️" as const,
  ur: "↗️" as const,
  dr: "↘️" as const,
  dl: "↙️" as const,
  ul: "↖️" as const,
};

export const emojiSquares = {
  black: "⬛" as const,
  white: "⬜" as const,
  yellow: "🟨" as const,
  green: "🟩" as const,
  red: "🟥" as const,
  purple: "🟪" as const,
};
export const fiveSquaresRegex = "(?:⬛|⬜|🟨|🟩){5}";

export const emojiCircles = {
  white: "⚪" as const,
  yellow: "🟡" as const,
  green: "🟢" as const,
  red: "🔴" as const,
};

export const listToValueMap = (list: Array<string>) =>
  Object.fromEntries(list.map((key, idx) => [key, idx]));

const numbers = "0️⃣1️⃣2️⃣3️⃣4️⃣5️⃣6️⃣7️⃣8️⃣9️⃣";
export const numberEmoji = splitEmoji(numbers);
export const numberEmojiRegex = emojiToRegexUnion(numberEmoji);
export const numberEmojiValues = listToValueMap(numberEmoji);
export function emojiToNumber(emoji: Array<string>) {
  return emoji.reduce((sum, numEmoji) => 10 * sum + numberEmojiValues[numEmoji], 0);
}

export function numberToEmoji(number: number, padZeros = 0) {
  if (number === 0) {
    return numberEmoji[0].repeat(Math.max(padZeros, 1));
  }
  const result: Array<string> = [];
  while (number > 0) {
    result.push(numberEmoji[number % 10]);
    number = Math.floor(number / 10);
  }
  const numZeros = padZeros - result.length;
  if (numZeros > 0) {
    result.push(numberEmoji[0].repeat(numZeros));
  }
  return result.reverse().join("");
}
