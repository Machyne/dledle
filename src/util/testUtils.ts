import { BaseGame, GameScore } from "../baseGame";

export type TestCases = {
  name: string;
  input: string;
  expectedScore: GameScore;
  expectedOutput?: string;
};

export function runGameTests(game: BaseGame, testCases: Array<TestCases>) {
  const winCase = testCases.find(({ expectedScore }) => expectedScore === GameScore.Win);
  if (winCase) {
    testCases.push({
      name: winCase.name + " with extra link text",
      input: "whatever text\n" + winCase.input + "\n" + game.link,
      expectedScore: GameScore.Win,
      expectedOutput: winCase.expectedOutput ?? winCase.input.trim(),
    });
    testCases.push({
      ...winCase,
      name: winCase.name + " trimmed",
      input: winCase.input.trim(),
    });
  }

  for (const { name, input, expectedScore, expectedOutput } of testCases) {
    it(`should serialize and deserialize ${name} correctly`, () => {
      const match = game.resultRegex.exec(input);
      expect(match).not.toBeNull();
      const { score, serializedResult } = game.serializeResult(match!);

      expect(score).toEqual(expectedScore);
      expect(game.deserialize(serializedResult)).toEqual(expectedOutput ?? input.trim());
    });
  }

  it("serialized data should match snapshot", () => {
    const data = Object.fromEntries(
      testCases.map(({ name, input }) => {
        const match = game.resultRegex.exec(input);
        expect(match).not.toBeNull();
        return [name, game.serializeResult(match!)];
      }),
    );
    expect(data).toMatchSnapshot();
  });
}
