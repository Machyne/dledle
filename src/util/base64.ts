import { DateParts } from "./dateHelpers";

// Feels like a good compromise between density and usability.
// 64^3 = 262,144 which is bigger than 3^10 = 59,049 or 101^2 = 10,201 so it can fit
// two rows of five 3-option guesses or two rows of 0-100 scores.
export const DEFAULT_WORD_WIDTH = 3;

// JS numbers are floats, with 52 bits of mantissa. This means we can represent ints with full precision up to 52 bits long (inclusive).
// This is plenty and it's easier to do it this way; no need to use BigInts.
// Math.floor(JS_MANTISSA_BITS / Math.log2(64)) = ⌊52/6⌋ = 8
const MAX_WORD_WIDTH = 8;

// From https://datatracker.ietf.org/doc/html/rfc4648#section-5
// Table 2: The "URL and Filename safe" Base 64 Alphabet
const B64_ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

const maxNumber = (wordWidth: number) => (1 << (6 * wordWidth)) - 1;

export function intToBase64(n: number, wordWidth = DEFAULT_WORD_WIDTH) {
  if (wordWidth > MAX_WORD_WIDTH) {
    throw new Error(`Word width ${wordWidth} is too large.`);
  }
  if (n < 0) {
    throw new Error(`${n} is negative.`);
  }
  if (n > maxNumber(wordWidth)) {
    throw new Error(`${n} is too large to fit in ${wordWidth} b64 characters.`);
  }

  let result = "";
  for (let i = 0; i < wordWidth; i++) {
    const b64Index = n % 64;
    n = Math.floor(n / 64);
    result = result + B64_ALPHABET[b64Index];
  }
  return result;
}

export function base64ToInt(s: string, wordWidth = DEFAULT_WORD_WIDTH) {
  if (wordWidth > MAX_WORD_WIDTH) {
    throw new Error(`Word width ${wordWidth} is too large.`);
  }
  if (s.length < wordWidth) {
    throw new Error(`${s} is not ${wordWidth} characters long.`);
  }
  let result = 0;
  for (let i = 0; i < wordWidth; i++) {
    var b64Index = B64_ALPHABET.indexOf(s[i]);
    if (b64Index === -1) {
      throw new Error(`${s[i]} is not a valid base64 character.`);
    }
    result += b64Index * 64 ** i;
  }
  return result;
}

export function nextInt(s: string, wordWidth = DEFAULT_WORD_WIDTH): [number, string] {
  if (s.length < wordWidth) {
    throw new Error(`${s} is not ${wordWidth} characters long.`);
  }
  const result = base64ToInt(s, wordWidth);
  return [result, s.slice(wordWidth)];
}

const startYear = 1990;
const maxYear = startYear + Math.floor(maxNumber(DEFAULT_WORD_WIDTH) / (31 * 12)) - 1;

export function datePartsToBase64(d: DateParts) {
  if (d.year < startYear || d.year > maxYear) {
    throw new Error(`Invalid year: ${d.year}`);
  }
  return intToBase64((d.year - startYear) * 31 * 12 + (d.month - 1) * 31 + (d.day - 1));
}

export function base64ToDateParts(s: string): DateParts {
  let intResult = base64ToInt(s);
  const day = (intResult % 31) + 1;
  intResult = Math.floor(intResult / 31);
  const month = (intResult % 12) + 1;
  intResult = Math.floor(intResult / 12);
  const year = intResult + startYear;
  return { year, month, day };
}

export function nextDateParts(s: string): [DateParts, string] {
  if (s.length < DEFAULT_WORD_WIDTH) {
    throw new Error(`${s} is not ${DEFAULT_WORD_WIDTH} characters long.`);
  }
  const result = base64ToDateParts(s);
  return [result, s.slice(DEFAULT_WORD_WIDTH)];
}

const shortStartYear = 2022;
// This gives a max of 2064
const shortMaxYear =
  shortStartYear + Math.floor(maxNumber(DEFAULT_WORD_WIDTH) / 16 / (31 * 12)) - 1;

export function shortDateToBase64(d: DateParts, extra4Bits: number) {
  if (d.year < shortStartYear || d.year > shortMaxYear) {
    throw new Error(`Invalid year: ${d.year}`);
  }
  if (extra4Bits < 0 || extra4Bits > 15) {
    throw new Error(`Invalid extra 4 bits: ${extra4Bits}`);
  }
  const dateNum = (d.year - shortStartYear) * 31 * 12 + (d.month - 1) * 31 + (d.day - 1);
  return intToBase64(dateNum * 16 + extra4Bits, DEFAULT_WORD_WIDTH);
}

export function base64ToShortDate(s: string): [DateParts, number] {
  let intResult = base64ToInt(s);
  const extra4Bits = intResult % 16;
  intResult = Math.floor(intResult / 16);
  const day = (intResult % 31) + 1;
  intResult = Math.floor(intResult / 31);
  const month = (intResult % 12) + 1;
  intResult = Math.floor(intResult / 12);
  const year = intResult + shortStartYear;
  return [{ year, month, day }, extra4Bits];
}

export function nextShortDate(s: string): [DateParts, string, number] {
  if (s.length < DEFAULT_WORD_WIDTH) {
    throw new Error(`${s} is not ${DEFAULT_WORD_WIDTH} characters long.`);
  }
  const result = base64ToShortDate(s);
  return [result[0], s.slice(DEFAULT_WORD_WIDTH), result[1]];
}
