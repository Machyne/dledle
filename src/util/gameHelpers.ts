import { DEFAULT_WORD_WIDTH, intToBase64, nextInt } from "./base64";
import { emojiSquares } from "./emoji";

export function emojiSquareScaleToNumber(guessChars: Array<string>) {
  let result = 0;
  if (guessChars.length < 5) {
    throw new Error(`Invalid guess: ${guessChars}`);
  }
  for (const guessChar of guessChars) {
    if (guessChar === emojiSquares.green) {
      result += 2;
    } else if (guessChar === emojiSquares.yellow) {
      result += 1;
    }
  }
  return result;
}

export function numberToEmojiSquareScale(num: number, padChar: string = emojiSquares.black) {
  if (num < 0 || num > 10) {
    throw new Error(`Invalid number: ${num}`);
  }
  const numGreen = Math.floor(num / 2);
  const numYellow = num % 2;
  return (
    emojiSquares.green.repeat(numGreen) +
    emojiSquares.yellow.repeat(numYellow) +
    padChar.repeat(5 - (numGreen + numYellow))
  );
}

export function encodeEmoji(emoji: Array<string>, values: Record<string, number>): number {
  const base = Math.max(...Object.values(values)) + 1;
  return emoji.reduce((sum, numEmoji, idx) => sum + values[numEmoji] * base ** idx, 0);
}

export function decodeEmoji(
  encoded: number,
  lookup: Array<string>,
  length: number = -1,
): Array<string> {
  const base = lookup.length;
  const ret: Array<string> = [];
  while (encoded > 0) {
    ret.push(lookup[encoded % base]);
    encoded = Math.floor(encoded / base);
  }
  if (length > 0) {
    if (ret.length > length) {
      throw new Error(`Encoded emoji too long: ${ret}`);
    }
    while (ret.length < length) {
      ret.push(lookup[0]);
    }
  }
  return ret;
}

export function emojiRowsToB64(args: {
  emojiLines: Array<Array<string>>;
  values: Record<string, number>;
  rowsPerB64Word: number;
  b64WordWidth?: number;
}): string {
  const rows = args.emojiLines;
  if (!rows.length) {
    return "";
  }
  const base = Math.max(...Object.values(args.values)) + 1;
  const b64WordWidth = args.b64WordWidth ?? DEFAULT_WORD_WIDTH;
  const rowOffset = base ** rows[0].length;

  if (args.rowsPerB64Word * rowOffset ** args.rowsPerB64Word > 64 ** b64WordWidth) {
    throw new Error(
      `Cannot encode ${args.rowsPerB64Word} rows of ${rows[0].length} emoji (base ${base}) into ${b64WordWidth} b64 chars`,
    );
  }
  let encodedStr = "";
  let currentCode = 0;
  for (let rowIdx = 0; rowIdx < rows.length; ++rowIdx) {
    const positionInRow = rowIdx % args.rowsPerB64Word;
    const isRowEnd = positionInRow === args.rowsPerB64Word - 1;
    const row = rows[rowIdx];
    const rowValue = encodeEmoji(row, args.values);
    currentCode += rowValue * rowOffset ** positionInRow;
    if (isRowEnd || rowIdx === rows.length - 1) {
      // Encode the number of rows in this block.
      currentCode = currentCode * args.rowsPerB64Word + positionInRow;
      encodedStr += intToBase64(currentCode, b64WordWidth);
      currentCode = 0;
    }
  }
  return encodedStr;
}

export function b64ToEmojiRows(args: {
  serializedResult: string;
  lookup: Array<string>;
  rowLength: number;
  rowsPerB64Word: number;
  b64WordWidth?: number;
}): { rows: Array<string>; serializedResult: string } {
  const b64WordWidth = args.b64WordWidth ?? DEFAULT_WORD_WIDTH;
  const base = args.lookup.length;
  const rowOffset = base ** args.rowLength;

  const rows: string[] = [];
  let intResult: number;
  let serializedResult = args.serializedResult;
  while (serializedResult.length > 0) {
    [intResult, serializedResult] = nextInt(serializedResult, b64WordWidth);
    const numRows = (intResult % args.rowsPerB64Word) + 1;
    intResult = Math.floor(intResult / args.rowsPerB64Word);
    for (let rowIdx = 0; rowIdx < numRows; ++rowIdx) {
      rows.push(decodeEmoji(intResult % rowOffset, args.lookup, args.rowLength).join(""));
      intResult = Math.floor(intResult / rowOffset);
    }
  }
  return { rows, serializedResult };
}
