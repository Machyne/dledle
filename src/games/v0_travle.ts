import { BaseGame, GameScore } from "../baseGame";
import { intToBase64, nextInt } from "../util/base64";
import { emojiSquares, emojiToRegexUnion, listToValueMap, splitEmoji } from "../util/emoji";
import { decodeEmoji, encodeEmoji } from "../util/gameHelpers";

const validHints = [emojiSquares.black, emojiSquares.red, emojiSquares.orange, "âœ…"];
const hintValues = listToValueMap(validHints);
const chunkSize = 3; // 2 bits per hint = 3 hints per base64 character.

export class V0Travle extends BaseGame {
  get name(): string {
    return "Travle";
  }

  get link(): string {
    return "https://travle.earth";
  }

  _buildResultRegex(): RegExp {
    const hintRe = emojiToRegexUnion(validHints);
    const optionalAway = "\\s*(?:\\((\\d+) away\\))?";

    return new RegExp(
      `#travle #(\\d+) \\(([0-9\\?]+)/(\\d+)\\)${optionalAway}\\s*(${hintRe}+)`,
      "u",
    );
  }

  serializeResult(gameResult: RegExpMatchArray) {
    const [, gameNumber, guessesTakenStr, maxGuessesStr, optionalAwayStr, guessesBlock] =
      gameResult;
    // Result will be: 2chars game number, 1char guesses taken, 1 char max guesses, optional 1 char away, and variable length guesses.
    let encoded = "";
    encoded += intToBase64(parseInt(gameNumber), 2);
    const guessesTaken = guessesTakenStr === "?" ? 0 : parseInt(guessesTakenStr);
    encoded += intToBase64(guessesTaken, 1);
    const maxGuesses = parseInt(maxGuessesStr);
    encoded += intToBase64(maxGuesses, 1);
    let optionalAway = 0;
    if (!!optionalAwayStr !== (guessesTaken === 0)) {
      throw new Error(`Invalid Travle away: ${optionalAwayStr} (guessesTaken: ${guessesTakenStr})`);
    }
    if (optionalAwayStr) {
      optionalAway = parseInt(optionalAwayStr);
      encoded += intToBase64(optionalAway, 1);
    }
    const guesses = splitEmoji(guessesBlock);
    const expectedNumGuesses = guessesTaken === 0 ? maxGuesses : guessesTaken;
    if (guesses.length !== expectedNumGuesses) {
      throw new Error(`Invalid Travle guesses: ${guessesBlock} (expected ${expectedNumGuesses})`);
    }
    while (guesses.length > 0) {
      const chunk = guesses.splice(0, chunkSize);
      encoded += intToBase64(encodeEmoji(chunk, hintValues), 1);
    }
    return {
      serializedResult: encoded,
      score:
        guessesTaken !== 0
          ? GameScore.Win
          : optionalAway === 1
            ? GameScore.NearWin
            : GameScore.Loss,
    };
  }

  deserialize(serializedResult: string) {
    if (serializedResult.length < 4) {
      throw new Error(`Invalid Travle result: ${serializedResult}`);
    }
    let intResult: number;
    [intResult, serializedResult] = nextInt(serializedResult, 2);
    const gameNumber = intResult;
    [intResult, serializedResult] = nextInt(serializedResult, 1);
    const guessesTaken = intResult;
    [intResult, serializedResult] = nextInt(serializedResult, 1);
    const maxGuesses = intResult;
    let optionalAway = 0;
    if (guessesTaken === 0) {
      [intResult, serializedResult] = nextInt(serializedResult, 1);
      optionalAway = intResult;
    }
    const expectedNumGuesses = guessesTaken === 0 ? maxGuesses : guessesTaken;
    const guesses: Array<string> = [];
    while (serializedResult.length > 0) {
      [intResult, serializedResult] = nextInt(serializedResult, 1);
      guesses.push(...decodeEmoji(intResult, validHints, chunkSize));
    }
    if (guesses.length < expectedNumGuesses) {
      throw new Error(`Invalid Travle guesses: ${guesses} (expected ${expectedNumGuesses})`);
    }
    guesses.splice(expectedNumGuesses);
    return `#travle #${gameNumber} (${
      guessesTaken === 0 ? "?" : guessesTaken.toString()
    }/${maxGuesses})${optionalAway === 0 ? "" : ` (${optionalAway} away)`}\n${guesses.join("")}`;
  }
}
