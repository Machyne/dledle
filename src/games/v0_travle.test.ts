import { GameScore } from "../baseGame";
import { runGameTests } from "../util/testUtils";
import { V0Travle } from "./v0_travle";

const sampleWins = [
  `
#travle #432 (6/10)
âœ…âœ…ðŸŸ§âœ…âœ…âœ…
`,
  `
#travle #409 (16/18)
ðŸŸ§âœ…ðŸŸ§âœ…âœ…ðŸŸ§âœ…âœ…âœ…ðŸŸ§âœ…ðŸŸ§âœ…âœ…âœ…âœ…
`,
];

const sampleLoss = `
#travle #432 (?/10) (2 away)
ðŸŸ§ðŸŸ§ðŸŸ§ðŸŸ¥ðŸŸ§ðŸŸ¥ðŸŸ§âœ…âœ…âœ…
`;

const sampleNearWin = `
#travle #432 (?/10) (1 away)
â¬›âœ…ðŸŸ¥ðŸŸ§ðŸŸ¥âœ…âœ…âœ…ðŸŸ§ðŸŸ§
`;

describe("V0Travle", () => {
  runGameTests(new V0Travle(), [
    ...sampleWins.map((input, idx) => ({
      name: `sampleWin${idx}`,
      input,
      expectedScore: GameScore.Win,
    })),
    {
      name: "sampleLoss",
      input: sampleLoss,
      expectedScore: GameScore.Loss,
    },
    {
      name: "sampleNearWin",
      input: sampleNearWin,
      expectedScore: GameScore.NearWin,
    },
  ]);
});
