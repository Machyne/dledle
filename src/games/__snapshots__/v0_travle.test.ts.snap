// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`V0Travle serialized data should match snapshot 1`] = `
{
  "sampleLoss": {
    "score": "Loss",
    "serializedResult": "wGAKCqZ-D",
  },
  "sampleNearWin": {
    "score": "NearWin",
    "serializedResult": "wGAKBc2vC",
  },
  "sampleWin0": {
    "score": "Win",
    "serializedResult": "wGGKv_",
  },
  "sampleWin0 trimmed": {
    "score": "Win",
    "serializedResult": "wGGKv_",
  },
  "sampleWin0 with extra link text": {
    "score": "Win",
    "serializedResult": "wGGKv_",
  },
  "sampleWin1": {
    "score": "Win",
    "serializedResult": "ZGQSuv_u_D",
  },
}
`;
